docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

```diff
@@ -0,0 +1,1680 @@
+              with
+                x2 : Optional Text
+                x1 : Int
+              controller p
+              do 
+                return ()
+
+Below, the choice on the right is **not** a valid upgrade of the choice
+on the left because it adds a new field ``x2`` before ``x1`` instead of adding
+it at the end of the parameter sequence.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+            choice C : ()
+              with
+                x1 : Int
+              controller p
+              do 
+                return ()
+
+     - .. code-block:: daml
+
+            choice C : ()
+              with
+                x2 : Optional Text
+                x1 : Int
+              controller p
+              do 
+                return ()
+
+Below, the choice on the right is **not** a valid upgrade of the choice
+on the left because it drops parameter ``x1``.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+            choice C : ()
+              with
+                x1 : Int
+              controller p
+              do 
+                return ()
+
+     - .. code-block:: daml
+
+            choice C : ()
+              with
+              controller p
+              do 
+                return ()
+
+Below, the choice on the right is **not** a valid upgrade of the choice
+on the left because it changes the type of ``x1`` from ``Int`` to ``Text``. ``Text`` is
+not a valid upgrade of ``Int``.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+            choice C : ()
+              with
+                x1 : Int
+              controller p
+              do 
+                return ()
+
+     - .. code-block:: daml
+
+            choice C : ()
+              with
+              controller p
+              do 
+                return ()
+
+Template Choices - Return Type
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The return type of an upgrading choice must be a valid upgrade of the
+return type of the upgraded choice.
+
+Changing the return type of a choice for a non-valid upgrade leads to a
+validation error.
+
+.. _examples-5:
+
+**Examples**
+
+Below, the choice on the right is **not** a valid upgrade of the choice
+on the left because it changes its return type from ``()`` to ``Int``. ``Int`` is
+not a valid upgrade of ``()``.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: daml
+
+            choice C : ()
+              controller p
+              do
+                return ()
+
+     - .. code-block:: daml
+
+            choice C : Int
+              controller p
+              do
+                return 1
+
+Data Types
+~~~~~~~~~~
+
+The serializable data types of the upgrading module must form a superset
+of the serializable data types of the upgraded package. In other words,
```

Comment:
Technical question: how do I know whether I'm declaring a serializable datatype or not?

Reply:
We want to have a better story for that for 2.10/3.1: a way to annotate types for serializability (or non-serializability maybe). But right now you have to rely on the LF spec to know whether your type is serializable or not, and you can't force it to be non-serializable.